test/SILOptimizer/licm_multiend.sil

```diff
@@ -0,0 +1,121 @@
+
+sil_stage canonical
+
+import Builtin
+import Swift
+import SwiftShims
+
+var x: Int
+
+let reversedArray: ReversedCollection<[Int]>
+
+// x
+sil_global hidden @$S3tmp1xSivp : $Int
+
+// reversedArray
+sil_global hidden [let] @$S3tmp13reversedArrays18ReversedCollectionVySaySiGGvp : $ReversedCollection<Array<Int>>
+
+// _swiftEmptyArrayStorage
+sil_global @_swiftEmptyArrayStorage : $_SwiftEmptyArrayStorage
+
+
+// CHECK-LABEL: sil hidden @multi_end_licm : $@convention(thin) () -> () {
+// CHECK: bb2:
+// CHECK: [[GLOBALVAR:%.*]] = global_addr @$S3tmp1xSivp : $*Int
+// CHECK: [[BEGINA:%.*]] = begin_access [modify] [dynamic] [no_nested_conflict] [[GLOBALVAR]] : $*Int
+// CHECK-NEXT: br bb4({{.*}} : $Builtin.Int64)
+// CHECK: bb4({{.*}} : $Builtin.Int64)
+// CHECK: cond_br {{.*}}, bb5, bb7
+// CHECK: bb5:
+// CHECK-NEXT: store
+// CHECK-NEXT: cond_br {{.*}}, bb6, bb9
+// CHECK: bb6:
+// CHECK-NEXT: end_access [[BEGINA]] : $*Int
+// CHECK-NEXT: bb10
+// CHECK: bb7:
+// CHECK-NEXT: struct $Int
+// CHECK-NEXT: store
+// CHECK-NEXT: cond_br {{.*}}, bb8, bb9
+// CHECK: bb8:
+// CHECK-NEXT: end_access [[BEGINA]] : $*Int
+// CHECK-NEXT: bb10
+// CHECK: bb9:
+// CHECK-NEXT: br bb4
+// CHECK: bb10:
+// CHECK-NEXT: br bb3
+sil hidden @multi_end_licm : $@convention(thin) () -> () {
+bb0:
+  %0 = global_addr @$S3tmp13reversedArrays18ReversedCollectionVySaySiGGvp : $*ReversedCollection<Array<Int>>
+  %1 = struct_element_addr %0 : $*ReversedCollection<Array<Int>>, #ReversedCollection._base
+  %2 = struct_element_addr %1 : $*Array<Int>, #Array._buffer
+  %3 = struct_element_addr %2 : $*_ArrayBuffer<Int>, #_ArrayBuffer._storage
+  %4 = struct_element_addr %3 : $*_BridgeStorage<_ContiguousArrayStorageBase, _NSArrayCore>, #_BridgeStorage.rawValue
+  %5 = load %4 : $*Builtin.BridgeObject
+  %6 = unchecked_ref_cast %5 : $Builtin.BridgeObject to $_ContiguousArrayStorageBase
+  %7 = ref_element_addr %6 : $_ContiguousArrayStorageBase, #_ContiguousArrayStorageBase.countAndCapacity
+  %8 = struct_element_addr %7 : $*_ArrayBody, #_ArrayBody._storage
+  %9 = struct_element_addr %8 : $*_SwiftArrayBodyStorage, #_SwiftArrayBodyStorage.count
+  %10 = struct_element_addr %9 : $*Int, #Int._value
+  %11 = load %10 : $*Builtin.Int64
+  %12 = builtin "assumeNonNegative_Int64"(%11 : $Builtin.Int64) : $Builtin.Int64
+  %13 = integer_literal $Builtin.Int64, 0
+  %14 = integer_literal $Builtin.Int1, 0
+  %15 = builtin "cmp_eq_Int64"(%12 : $Builtin.Int64, %13 : $Builtin.Int64) : $Builtin.Int1
+  %16 = builtin "int_expect_Int1"(%15 : $Builtin.Int1, %14 : $Builtin.Int1) : $Builtin.Int1
+  cond_br %16, bb1, bb2
+
+bb1:
+  br bb3
+
+bb2:
+  %19 = global_addr @$S3tmp1xSivp : $*Int
+  %20 = integer_literal $Builtin.Int64, 1
+  %21 = integer_literal $Builtin.Int1, -1
+  %23 = ref_tail_addr %6 : $_ContiguousArrayStorageBase, $Int
+  br bb4(%12 : $Builtin.Int64)
+
+bb3:
+  %25 = tuple ()
+  return %25 : $()
```

Comment:
I think it would help if you layout the return block at the end of the function.